src/puya/awst/nodes.py

```diff
@@ -389,13 +389,51 @@ def accept(self, visitor: ExpressionVisitor[T]) -> T:
 
 @attrs.frozen
 class ARC4ArrayEncode(Expression):
-    values: Sequence[Expression]
     wtype: wtypes.ARC4StaticArray | wtypes.ARC4DynamicArray = attrs.field()
+    values: Sequence[Expression] = attrs.field(default=(), converter=tuple[Expression, ...])
 
     def accept(self, visitor: ExpressionVisitor[T]) -> T:
         return visitor.visit_arc4_array_encode(self)
 
 
+@attrs.frozen
+class ArrayConcat(Expression):
+    """
+    Given 'left' or 'right' that is logically an array - concat it with the other value which is
+    an iterable type with the same element type
+    """
+
+    left: Expression
+    wtype: wtypes.WType
+    right: Expression
+
+    def accept(self, visitor: ExpressionVisitor[T]) -> T:
+        return visitor.visit_array_concat(self)
+
+
+@attrs.frozen
+class ArrayPop(Expression):
+    base: Expression
+
+    def accept(self, visitor: ExpressionVisitor[T]) -> T:
+        return visitor.visit_array_pop(self)
+
+
+@attrs.frozen
+class ArrayExtend(Expression):
```

Comment:
I think this might be better as a statement rather than an Expression with a void wtype? As it is analogous to the AugmentedAssignment nodes

Reply:
I did try that however it would require a change to the expression builder to allow statements to be returned from calls. This felt like the path of least resistance.